File: Challenges/LeetCode/DesignAddAndSearchWordsDataStructure.go
type WordDictionary struct {
	root *WordNode
}

type WordNode struct {
	Ending   bool
	children map[rune]*WordNode
}

func (wn *WordNode) GetOrAdd(r rune) *WordNode {
	var node *WordNode
	node, ok := wn.children[r]
	if !ok {
		node = &WordNode{
			children: make(map[rune]*WordNode),
		}
		wn.children[r] = node
	}

	return wn.children[r]
}

func Constructor() WordDictionary {
	return WordDictionary{
		root: &WordNode{
			children: make(map[rune]*WordNode),
		},
	}
}

func (this *WordDictionary) AddWord(word string) {
	curr := this.root
	for ix, r := range word {
		next := curr.GetOrAdd(r)
		curr = next
		if ix == len(word)-1 {
			next.Ending = true
		}
	}
}

func (this *WordDictionary) Search(word string) bool {
	w := []rune(word)

	type state struct {
		ix int
		wn *WordNode
	}
	queue := make([]state, 0)
	queue = append(queue, state{ix: 0, wn: this.root})
	for len(queue) > 0 {
		curr := queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		if curr.ix == len(w) {
			if curr.wn.Ending {
				return true
			}
			continue
		}

		r := w[curr.ix]
		if r == '.' {
			for _, c := range curr.wn.children {
				queue = append(queue, state{ix: curr.ix + 1, wn: c})
			}
		} else if c, ok := curr.wn.children[r]; ok {
			queue = append(queue, state{ix: curr.ix + 1, wn: c})
		}
	}

	return false
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddWord(word);
 * param_2 := obj.Search(word);
 */

File: Challenges/LeetCode/DesignAddAndSearchWordsDataStructure.cs
using Challenges.LeetCode.ConstructQuadTree;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Challenges.LeetCode.DesignAddAndSearchWordsDataStructure
{
    public class WordDictionary
    {
        public class Node
        {
            public Dictionary<char, Node> Children { get; } = new Dictionary<char, Node>();
            public char C { get; }
            public int Depth { get; }
            public bool Word { get; set; }


            public Node(char c, int depth)
            {
                C = c;
                Depth = depth;
            }
        }


        public Node Head { get; private set; }


        public WordDictionary()
        {
            Head = new Node('\0', 0);
        }

        public void AddWord(string word)
        {
            Node curr = Head;
            for (int i = 0; i < word.Length; i++)
            {
                char c = word[i];
                if (!curr.Children.ContainsKey(c))
                {
                    curr.Children.Add(c, new Node(c, i + 1));
                }
                curr = curr.Children[c];
                if (i == word.Length - 1)
                {
                    curr.Word = true;
                }
            }
        }

        public bool Search(string word)
        {
            var queue = new Queue<Node>();
            queue.Enqueue(Head);
            while (queue.Any())
            {
                Node node = queue.Dequeue();
                if (word.Length == node.Depth)
                {
                    if (node.Word)
                        return true;
                }
                else
                {
                    char c = word[node.Depth];
                    if (c == '.')
                    {
                        foreach (Node child in node.Children.Values)
                            queue.Enqueue(child);
                    }
                    else if (node.Children.ContainsKey(c))
                    {
                        queue.Enqueue(node.Children[c]);
                    }
                }
            }
            return false;
        }
    }

    /**
     * Your WordDictionary object will be instantiated and called as such:
     * WordDictionary obj = new WordDictionary();
     * obj.AddWord(word);
     * bool param_2 = obj.Search(word);
     */
}
